libazureinit/src/provision/mod.rs

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+pub mod hostname;
+pub mod password;
+pub mod user;
+
+use std::{
+    fs::Permissions,
+    io::Write,
+    marker::PhantomData,
+    os::unix::fs::{DirBuilderExt, PermissionsExt},
+};
+
+use private::*;
+use tracing::instrument;
+
+use crate::{error::Error, imds::PublicKeys};
+
+// State transitions for the Provision structure which are used to force
+// the user to provide backends for things they wish to provision.
+mod private {
+    #[derive(Default)]
+    pub struct NoBackends;
+    #[derive(Default)]
+    pub struct HostnameBackend;
+    #[derive(Default)]
+    pub struct PasswordBackend;
+    #[derive(Default)]
+    pub struct AllBackends;
+}
+
+#[derive(Default)]
+pub struct Provision<State = NoBackends> {
+    hostname: String,
+    username: String,
+    keys: Vec<PublicKeys>,
+    password: Option<String>,
+    hostname_backends: Vec<hostname::Provisioner>,
+    user_backends: Vec<user::Provisioner>,
+    password_backends: Vec<password::Provisioner>,
+    state: PhantomData<State>,
+}
+
+impl Provision<NoBackends> {
+    pub fn new(hostname: String, username: String) -> Self {
+        Self {
+            hostname,
+            username,
+            ..Default::default()
+        }
+    }
+
+    pub fn hostname_provisioners(
+        mut self,
+        backends: impl Into<Vec<hostname::Provisioner>>,
+    ) -> Provision<HostnameBackend> {
+        self.hostname_backends = backends.into();
+        Provision {
+            hostname: self.hostname,
+            username: self.username,
+            keys: self.keys,
+            password: self.password,
+            hostname_backends: self.hostname_backends,
+            user_backends: self.user_backends,
+            password_backends: self.password_backends,
+            state: PhantomData,
+        }
+    }
+}
+
+impl Provision<HostnameBackend> {
+    pub fn user_provisioners(
+        mut self,
```

Comment:
Should `user_provisioners` be located under `impl Provisioners<UseraddBackend>`?

Reply:
So this is one of the things I didn't really like about my approach - some arguments (user, hostname) are currently required as they're part of `new`, whereas some arguments (currently just password) are not necessarily required. So the state transitions are a little weird as a consequence of that + not having a default for the backends. We could make the hostname and username behave like password so usage would be like:

```
Provision::new()
    .hostname("my-host")
    .hostname_backends(...)
    .user("my-user")
    .username_backends(...)
    .password("my-password")
    .password_backends(...)
    .provision();
```

Also, I think the above interface (and what I've currently implemented) only really makes sense if you're not required to specify a backend, in which case the state transitions are kind of pointless anyway. This would be a simpler interface if we required explicitly selecting backends:

```
Provision::new()
    .hostname("my-host", &[backend1, backend2,])
    .user("my-user", &[backend1, backend2])
    .password("my-password", &[backend1, backend2])
    .provision();
```

I go back and forth on having the user be explicit about how to do the thing. One thing I don't like about what I've currently got is there's no interface for users to say "do it all, and any new backends the library implements in the future". That being the default makes it possible, but we could also add a thing for the user to explicitly opt-in to that. What do you think?

Reply:
Oh, thanks for the detailed explanation.
If I would have to choose between the 2 examples, i.e. complicated interface and a simpler interface, then I would choose the latter, a simpler one.
I have no better idea at the moment.

Reply:
Okay, I decided to go back to the "if you don't provide a specific list of backends we'll try everything" approach, so a minimal use of the API would be:
```
# minimal usecase
Provision::new("my-host", "my-user", [pubkey1, ...]).provision();

# more specific
Provision::new("my-host", "my-user", [pubkey1, ...])
    .password("my-password")
    .hostname_backends([HostnameBackend::Hostnamectl])
    .provision();
```

I decided to make providing SSH keys mandatory (although you can provide an empty list if you really want).